pkg/controller/pubsub/pubsub_config_controller.go

```diff
@@ -0,0 +1,155 @@
+package pubsub
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+
+	constraintclient "github.com/open-policy-agent/frameworks/constraint/pkg/client"
+	"github.com/open-policy-agent/frameworks/constraint/pkg/externaldata"
+	"github.com/open-policy-agent/gatekeeper/pkg/expansion"
+	"github.com/open-policy-agent/gatekeeper/pkg/logging"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation"
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub"
+	psClient "github.com/open-policy-agent/gatekeeper/pkg/pubsub/provider"
+	"github.com/open-policy-agent/gatekeeper/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	"github.com/open-policy-agent/gatekeeper/pkg/watch"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var (
+	pubsubEnabled = flag.Bool("enable-pub-sub", false, "Enabled pubsub to publish messages")
+	log           = logf.Log.WithName("controller").WithValues(logging.Process, "pubsub_controller")
+)
+
+type Adder struct {
+	PubsubSystem *pubsub.System
+}
+
+func (a *Adder) Add(mgr manager.Manager) error {
+	if !*pubsubEnabled {
+		return nil
+	}
+	r := newReconciler(mgr, a.PubsubSystem)
+	return add(mgr, r)
+}
+
+func (a *Adder) InjectOpa(_ *constraintclient.Client) {}
+
+func (a *Adder) InjectWatchManager(_ *watch.Manager) {}
+
+func (a *Adder) InjectControllerSwitch(_ *watch.ControllerSwitch) {}
+
+func (a *Adder) InjectTracker(_ *readiness.Tracker) {}
+
+func (a *Adder) InjectMutationSystem(_ *mutation.System) {}
+
+func (a *Adder) InjectExpansionSystem(_ *expansion.System) {}
+
+func (a *Adder) InjectProviderCache(_ *externaldata.ProviderCache) {}
+
+func (a *Adder) InjectPubsubSystem(pubsubSystem *pubsub.System) {
+	a.PubsubSystem = pubsubSystem
+}
+
+type Reconciler struct {
+	client.Client
+	scheme *runtime.Scheme
+	system *pubsub.System
+}
+
+func newReconciler(mgr manager.Manager, system *pubsub.System) *Reconciler {
+	return &Reconciler{
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+		system: system,
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New("pusbub-config-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(
+		&source.Kind{Type: &corev1.ConfigMap{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
```

Comment:
yes, it makes sure to only reconcile for config maps for Gatekeeper's namespace.

Reply:
The problem isn't so much the reconcile as the watch... the informer cache (invoked by c.Watch()) will cache the contents of all config maps it receives an event for. This caching happens before the predicate functions are called.

Reply:
Yeah, true. I was looking into how constraint template CRDs are implemented. If everyone is on board, I can follow up with changes that introduce a similar mechanism for Pubsub. As I am understanding, I think we will need something similar right?

A new CRD - `pubsub_connection_template` similar to `constraint_template`
and `dapr_connection`, `rabbitmq_connection` of type `pubsub_connection_template`
and then different connections of type `dapr_conenction`

Reply:
I don't think we need anything as complex as constraint template's watching mechanism. Since users would need to fork the code to integrate new clients, hardcoded watches are fine (so long as they are cheap to implement).

IMO each pubsub could have its own kind and write its own controller (registering a controller in G8r wouldn't be that big of a fork), or we can define a kind that accepts arbitrary contents, which is where the connection config would go. Mutation does this:

https://github.com/open-policy-agent/gatekeeper/blob/a7d299e6cf67ef2437a70e7b1c87822d56617cb5/apis/mutations/v1/value.go#L12

Reply:
I don't have any strong opinions on how this should be implemented. What you are suggesting makes sense to me.

Reply:
Ack, the other option I could think of would be to use a single kind for all configs and just have pluggable controllers. That might save some boilerplate. Not obvious which is better. No technical reason we can't have both, though might be confusing to users.